From: Christoph Hellwig <hch@lst.de>
Date: Fri, 11 Mar 2016 17:34:51 +0100
Subject: block: factor out chained bio completion
Git-commit: 38f8baae890561203ba6093f76b14576ce9b271b
Patch-mainline: v4.6-rc1
References: FATE#319965, bsc#964944

Factor common code between bio_chain_endio and bio_endio into a common
helper.

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jens Axboe <axboe@fb.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.com>
---
 block/bio.c |   16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

--- a/block/bio.c
+++ b/block/bio.c
@@ -296,14 +296,19 @@ void bio_reset(struct bio *bio)
 }
 EXPORT_SYMBOL(bio_reset);
 
-static void bio_chain_endio(struct bio *bio)
+static struct bio *__bio_chain_endio(struct bio *bio)
 {
 	struct bio *parent = bio->bi_private;
 
 	if (!parent->bi_error)
 		parent->bi_error = bio->bi_error;
-	bio_endio(parent);
 	bio_put(bio);
+	return parent;
+}
+
+static void bio_chain_endio(struct bio *bio)
+{
+	bio_endio(__bio_chain_endio(bio));
 }
 
 /*
@@ -1758,12 +1763,7 @@ void bio_endio(struct bio *bio)
 		 * pointers also disables gcc's sibling call optimization.
 		 */
 		if (bio->bi_end_io == bio_chain_endio) {
-			struct bio *parent = bio->bi_private;
-
-			if (!parent->bi_error)
-				parent->bi_error = bio->bi_error;
-			bio_put(bio);
-			bio = parent;
+			bio = __bio_chain_endio(bio);
 		} else {
 			if (bio->bi_end_io)
 				bio->bi_end_io(bio);
